src/storage/src/healthcheck.rs

```diff
@@ -310,10 +308,9 @@ impl HealthOperator for DefaultWriter {
 /// A health message consumed by the `health_operator`.
 #[derive(Clone, Debug, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
 pub struct HealthStatusMessage {
-    /// The index of the object this message describes.
-    ///
-    /// Useful for sub-objects like sub-sources.
-    pub index: usize,
+    /// The object that this status message is about. When None, it refers to the entire ingestion
+    /// as a whole. When Some, it refers to a specific subsource.
+    pub id: Option<GlobalId>,
```

Comment:
The first set of changes relate to changing the healthcheck operator to identify statuses based on the `GlobalId` of the subsource instead of an index. The healthcheck operator has a special `GlobalId`, called the `halting_id`, which is the global id whose statuses are allowed to halt the source. Instead of having a magic id I chose to represent those special statuses as those having `id: None`, which can be thought of as a status update for the entire ingestion.